### PR TITLE
ICU-22968 Rearrange bits in trie values in normalization data export for ICU4X

### PR DESCRIPTION
**Pull request by hsivonen**
This changes the format of the normalization data export for ICU4X in the way expected by https://github.com/unicode-org/icu4x/pull/5813 , which makes the ICU4X normalizer significantly faster.

#### Checklist
- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22968
- [x] Required: The PR title must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by hsivonen** [comment]
The Jira issue is in the Accepted state, so checking the checkbox here.

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp [hunk 1]
```diff
@@ -749,6 +737,8 @@ size_t findOrAppend(std::vector<T>& storage, const UChar32* needle, size_t needl
 
 
 // Computes data for canonical decompositions
+// See components/normalizer/trie-value-format.md in the ICU4X repo
```
Comment (echeran) [inline icu4c/source/tools/icuexportdata/icuexportdata.cpp:740]: praise: thanks for adding this pointer to the documentation

**Review by echeran — DISMISSED** [review]
Looks good, thanks! I can re-approve and/or merge in case you need it.

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp [hunk 2]
```diff
@@ -591,28 +606,63 @@ void writeDecompositionData(const char* basename, uint32_t baseSize16, uint32_t
...
-        umutablecptrie_set(builder.getAlias(), pending.scalar, swappedTrieValue, status);
     }
+}
+
+/// Marker that the decomposition does not round trip via NFC.
+const uint32_t NON_ROUND_TRIP_MARKER = (1 << 30);
+
+/// Marker that the first character of the decomposition can combine
+/// backwards.
+const uint32_t BACKWARD_COMBINING_MARKER = (1 << 31);
```
[the remark this answers is not in the document]
Comment (hsivonen) [reply icu4c/source/tools/icuexportdata/icuexportdata.cpp]: Changed.

File: icu4c/source/tools/icuexportdata/icuexportdata.cpp [hunk 3]
```diff
@@ -489,12 +489,29 @@ FILE* prepareOutputFile(const char* basename) {
 
 #if !UCONFIG_NO_NORMALIZATION
 
-struct PendingDescriptor {
+class PendingDescriptor {
+public:
     UChar32 scalar;
     uint32_t descriptor;
```
[the remark this answers is not in the document]
Comment (hsivonen) [reply icu4c/source/tools/icuexportdata/icuexportdata.cpp]: Changed.

**Comment by hsivonen** [comment]
> Looks good, thanks! I can re-approve and/or merge in case you need it.

Thank you. I think this needs re-approval and merging now.

**Review by echeran — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**